### ISSUE TITLE
Buffering stdout/stderr from child process

### ISSUE THREAD

**Issue by ORESoftware** [nodejs/help#963]
Say I launch a child process with a reasonable amount of stdout/stderr:

  ```js
    const k = cp.spawn('bash');
    
    k.stdout.pause();
    k.stderr.pause();
    
    setTimeout(function(){
       // 20  seconds later, we resume streaming
       k.stdout.pipe(getDest());
       k.stderr.pipe(getDest());
    }, 20000);

```

with the stdout/stderr for this child process really get buffered in memory and will none of it be lost when I finally resume streaming (after 20 seconds and _even if_ the child process has exited already)?

Should this always work or are there some edge cases to take into consideration?

**Comment by ORESoftware** [comment]
https://stackoverflow.com/questions/47228540/buffering-stdout-stderr-from-child-process

still looking for a definitive answer to this

**Comment by gireeshpunathil** [comment]
cat foo.cc

```c
#include <stdio.h>
#include <unistd.h>

int main() {
  int i = 0;
  while(i++ < 10) {
    printf("%d\n", i);
    fflush(stdout);
    sleep(1);
  }
}
```

```js
const cp = require('child_process')
const fs = require('fs')
const str = fs.createWriteStream("./out.txt")
const k = cp.spawn('./a.out');
k.stdout.pause();
setTimeout(function(){
  k.stdout.pipe(str)
}, 2000);
```


```strace
[pid 20499] write(1, "1\n", 2 <unfinished ...>
[pid 20489] epoll_wait(3,  <unfinished ...>
[pid 20489] read(14,  <unfinished ...>
[pid 20489] <... read resumed> "1\n", 65536) = 2
[pid 20499] nanosleep({1, 0},  <unfinished ...>
[pid 20489] epoll_wait(3,  <unfinished ...>
[pid 20499] write(1, "2\n", 2)          = 2
[pid 20489] <... epoll_wait resumed> {{EPOLLIN, {u32=14, u64=14}}}, 1024, 1995) = 1
[pid 20489] read(14,  <unfinished ...>
[pid 20489] <... read resumed> "2\n", 65536) = 2
[pid 20499] nanosleep({1, 0},  <unfinished ...>
[pid 20489] epoll_wait(3, {}, 1024, 997) = 0
```
PID `20489` is the `parent` and `20499` is the `child`.

the stream.pause() is a higher level abstraction on the stream object, does not penetrate to the fd / network level. It's effect will be only visible on the stream interfaces with the program - no events will be emitted pertinent to dataflow.

Having said that, the scenario you illustrate does not provide any guarentee of data retention in the child or availability in the parent. It is a pure function of how much data is flowing, and how much data the parent's internal buffer is left with etc. For example, if the child writes megabytes of data while parent stream is paused, the internal buffer may stop reading when its highWaterMark has been reached. Also, if the child exits before the data could not be fully read by the network interface, the data will be lost. Finally, there are platform specific behaviors - for example some windows flavors inhibit process terminations before the I/O can be fully flushed out.

In short, the answer depends. In simple and most common cases, the data is available to the parent thrugh the internal buffering. Hope this explanation helps.

**Comment by gireeshpunathil** [comment]
closing as answered, please let me know if you have outstanding questions on this.
